### DISCUSSION TITLE
Table rows as structs

### DISCUSSION THREAD

**Discussion by mikekap** [duckdb/duckdb#3558] [General]
#### What happens?
You can't convert a table alias to a struct.

#### To Reproduce
```sql
WITH data AS (
SELECT 1 as a, 2 as b, 3 as c
)

SELECT d FROM data d
```
In most SQL dialects that support row types, this gives you a single column with a row type (i.e. struct). You can see it specifically mentioned in the [postgres docs](https://www.postgresql.org/docs/current/rowtypes.html#ROWTYPES-USAGE).

#### Environment (please complete the following information):
 - OS: OSX
 - DuckDB Version: `duckdb-0.3.5.dev167-cp38-cp38-macosx_10_9_x86_64.whl`
 - DuckDB Client: Python

#### Before Submitting

- [x] **Have you tried this on the latest `master` branch?**
* **Python**: `pip install duckdb --upgrade --pre`
* **R**: `install.packages("https://github.com/duckdb/duckdb/releases/download/master-builds/duckdb_r_src.tar.gz", repos = NULL)`
* **Other Platforms**: You can find binaries [here](https://github.com/duckdb/duckdb/releases/tag/master-builds) or compile from source.

- [x] **Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?**

**Comment by hannes** [comment]
Interesting, what already works though is 

```SQL
WITH data AS (
SELECT 1 as a, 2 as b, 3 as c
)

SELECT row(a,b,c) FROM data d;
```

**Reply by mikekap** [reply]
Right that definitely works, but requires listing all the columns. Broadly this is useful to get a `min/max-by` idiom. For example, this is how I would do a "latest row for each user" in bigquery SQL:

```sql
SELECT ARRAY_AGG(row ORDER BY timestamp DESC LIMIT 1)[SAFE_OFFSET(0)].*
FROM table row
GROUP BY row.user_id
```

duckdb supports the majority of that - and the only vital missing feature is table rows as structs.

**Reply by hannes** [reply]
Well, happy to review a PR!

**Reply by hannes** [reply]
@lnkuiper was quicker: #3569
